## Resumo de Amostragem 1 com código em Python

**prof. Davi Rocha**

### Amostra
Parte de uma população, selecionada usando alguma técnica que de chances iguais a todos os elementos da população de serem selecionados

Conceitos

- População: alvo do estudo
- Amostra: subconjunto da população
- Censo: pesquisa com toda população

### É possível inferir sobre uma amostra

- Uma amostra feita corretamente deve representar as mesmas características da população de onde foi retirada.

- Se ela não representa a população, dizemos que ele é enviesada

#### Enviesamento
Você subestima ou superestima o parâmetro da população

##### Causas:

- Pesquisa de pessoas próximas ou de fácil acesso

- Pesquisas pela Internet

- Sem uso de mecanismo de seleção aleatório

##### “Custo” da Amostra

- Margem de Erro e Nível de Confiança

- Variação: amostras diferentes podem apresentar resultados diferentes

- Podemos “medir” a variação esperada

### Amostragem

- Amplamente utilizada em pesquisas, estudos etc.

- Faz parte do dia a dia de um Cientista de Dados

#### Principais tipos de amostras

- Aleatória Simples
- Estratificada
- Sistemática

Aqui veremos só a AAS(Amostra Aleatória Simples), nos próximos resumos falaremos dos outros tipos

### Amostras Aleatórias Simples(AAS)
• Um determinado número de elementos é retirado da população de forma aleatória

• Todos os elementos da população alvo do processo de amostragem, devem ter as mesmas chances de serem selecionados para fazer parte da amostra

#### Amostras Aleatórias Simples(AAS):

- Com reposição - depois de selecionada a unidade amostral volta para a amostra e pode ser selecionada novamente em uma nova realização do experimento
- Sem reposição - depois de selecionada a unidade amostral não volta para a amostra

Vamos ver como fazer AAS no Python

#### Amostras Aleatórias Simples no Python(com reposição)

1) Vamos gerar 150 amostras de 0’s e 1’s, com reposição com probabilidade igual de ser escolhido(pode pensar como o lançamento de uma moeda, por exemplo 0=cara e 1=coroa)

Vamos importa o pandas e o numpy


In [1]:
import pandas as pd
import numpy as np

In [3]:
np.random.seed(2345) #semente aleatória
amostra = np.random.choice(a = [0, 1], size = 150, replace = True,  p = [0.5, 0.5])
#no R: amostra<- sample(x = c(0,1), size = 150,replace = TRUE, prob = c(0.5,0.5)) veja a semelhança!!!
amostra

array([1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0,
       0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1])

Contando de 0’s e 1’s:

In [13]:
#percentual de 0's:
round(sum(amostra==0)/len(amostra)*100,2)

46.67

In [14]:
#percentual de 1´s
round(sum(amostra==1)/len(amostra)*100,2)

53.33

Está muito próximo de 50% para cada! Se aumentarmos a amostra provavelmente ficarão mais próximo ainda de 50% de acordo com a lei dos grandes números.

2) Vamos gerar 150 amostras de 0’s e 1’s, com reposição com probabilidade maior para 0’s. Você pode pensar numa moeda enviesada

In [17]:
amostra_30 =np.random.choice([0,1], size=150, replace=True,p=[0.7,0.3])
amostra_30

array([0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [18]:
#percentual de 0's:
round(sum(amostra_30==0)/len(amostra_30)*100,2)

75.33

In [19]:
round(sum(amostra_30==1)/len(amostra_30)*100,2)

24.67

Está próximo de 70% - 30% como queriamos! Para chegar mais próximo ainda basta aumentar a amostra!!

#### Amostras Aleatórias Simples no Python(sem reposição)

3) Agora vamos selecionar 150 amostras, de 1 a 1000, sem reposição, com probabilidades equivalentes

In [25]:
amostra_sem_rep= np.random.choice(np.arange(1,1001), size=150, replace=False)
amostra_sem_rep

array([732, 319, 632,  28, 827,  76, 663, 659, 995, 942, 316, 190,  10,
       935, 536, 946, 706,  39, 531, 494, 985, 141, 885, 788, 185, 569,
       498, 800, 107, 451, 294, 315,  37,   8, 155, 460, 169, 831, 678,
       237, 852,  33, 916, 709, 809, 261, 257, 589, 833,  13,  66, 808,
       301, 638, 668, 652, 719, 235, 260, 650,  15, 104,  69, 990, 910,
       414, 671, 737, 622,  59, 305,  45, 443, 912, 400, 664, 751,  64,
       233,  92, 142, 470, 890, 140, 755, 907, 969, 584, 438, 768,  90,
       669, 361,  25, 375, 972, 807, 115, 983, 970, 646, 203,  22, 621,
       909, 692, 505, 501, 625, 227, 674, 324, 733, 884, 739, 956, 492,
       196, 446, 422, 924, 832, 343, 976, 762, 849,  85,  24, 861, 359,
       575, 859, 781, 881, 256, 429, 618, 631, 894,  36, 131, 892, 268,
       731, 722, 951, 518, 170, 856,  57])

In [23]:
np.arange(1,10)

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

4) Gerando amostra de conjuntos de dados - Vamos gerar uma amostra aleatória de 30% dos dados mpg



In [37]:
import seaborn as sns #para acessar os dados

In [40]:
mpg=sns.load_dataset('mpg')  #acessando os dados

In [41]:
#visualizando uma amostra de dados
mpg.sample(3)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
40   14.0          8         351.0       153.0    4154          13.5   
366  17.6          6         225.0        85.0    3465          16.6   
181  33.0          4          91.0        53.0    1795          17.5   

     model_year origin                    name  
40           71    usa        ford galaxie 500  
366          81    usa  chrysler lebaron salon  
181          75  japan        honda civic cvcc

In [46]:
mpg.shape

(398, 9)

Há 398 linhas e 9 colunas

Agora vamos separar esse conjunto em duas partes um com 30% dos dados e outro em 70% dos dados. Esse tipo de divisão é feita para treinar modelos de Machine Learning por exemplo.

Para isso primeiro vamos criar uma amostra com 70% de 0’s e 30% de 1’s como fizemos anteriormente:

In [47]:
amostra= np.random.choice(a=[0,1], size=mpg.shape[0], replace= True,p=[0.7,0.3])

Contagem:

In [52]:

sum(amostra==0), sum(amostra==1)

(268, 130)

Deixo para você se essa proporção está aproximadamente correta. Vamos agora dividir o conjunto mpg usando essas duas amostras. Reforço que você poderia fazer esse mesmo processo para separar um conjunto em treino e teste para aplicar um algoritmo de Machine Learning! Vamos lá:

In [54]:
mpg_70=mpg[amostra==0]
mpg_30=mpg[amostra==1]

In [57]:
mpg_70.shape[0]

268

In [59]:
mpg_70.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
4  17.0          8         302.0       140.0    3449          10.5   
5  15.0          8         429.0       198.0    4341          10.0   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
4          70    usa                ford torino  
5          70    usa           ford galaxie 500

In [58]:
mpg_30.shape[0]

130

In [61]:
mpg_30.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
3   16.0          8         304.0       150.0    3433          12.0   
6   14.0          8         454.0       220.0    4354           9.0   
7   14.0          8         440.0       215.0    4312           8.5   
8   14.0          8         455.0       225.0    4425          10.0   
10  15.0          8         383.0       170.0    3563          10.0   

    model_year origin                 name  
3           70    usa        amc rebel sst  
6           70    usa     chevrolet impala  
7           70    usa    plymouth fury iii  
8           70    usa     pontiac catalina  
10          70    usa  dodge challenger se

Verificando a proporção das amostras:

In [65]:
#Proporção de 0's
mpg_70.shape[0]/len(mpg)

0.6733668341708543

In [66]:
#Proporção de 1's
mpg_30.shape[0]/len(mpg)

0.32663316582914576

Portanto dividimos o dataset mpg em duas partes escolhidas de forma aleatória com aproximadamente 30% e 70% respectivamente.

Era isso que eu queria mostrar.

**Keep calm and analysing data!**

